File: mobile/api/models/studentModel.js
import pool from '../config/db.js';

export const createStudent = async (studentData) => {
  const {
    student_name,
    registration_number,
    date_of_birth,
    gender,
    address,
    father_name,
    mother_name,
    assigned_class,
    assigned_section,
    birthCertificatePath,
    studentPhotoPath,
    signup_id,
    session_id
  } = studentData;

  const sql = `
    INSERT INTO students (
      student_name,
      registration_number,
      date_of_birth,
      gender,
      address,
      father_name,
      mother_name,
      assigned_class,
      assigned_section,
      birth_certificate,
      student_photo,
      signup_id,
      session_id
    ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)
    RETURNING *
  `;

  try {
    const result = await pool.query(sql, [
      student_name,
      registration_number,
      date_of_birth,
      gender,
      address,
      father_name,
      mother_name,
      assigned_class,
      assigned_section,
      birthCertificatePath,
      studentPhotoPath,
      signup_id,
      session_id
    ]);
    return result.rows[0];
  } catch (err) {
    console.error('Error creating student:', err);
    throw err;
  }
};

// Parent ko email se find karo signup table me
export const findSignupByEmail = async (email) => {
  const res = await pool.query('SELECT * FROM signup WHERE email = $1', [email]);
  return res.rows[0];
};

// Parent-Student link insert karo
export const linkParentStudent = async (parentSignupId, studentId) => {
  const sql = `INSERT INTO parent_student_link (parent_signup_id, student_id) VALUES ($1, $2) ON CONFLICT DO NOTHING`;
  await pool.query(sql, [parentSignupId, studentId]);
};

export const getStudentsByUser = async (signup_id) => {
  const query = `
    SELECT s.*
    FROM students s
    JOIN signup u ON s.signup_id = u.id
    WHERE u.school_id = (
      SELECT school_id FROM signup WHERE id = $1
    )
  `;

  try {
    const result = await pool.query(query, [signup_id]);
    return result.rows;
  } catch (err) {
    console.error('Error fetching students by user (school):', err);
    throw err;
  }
};

export const getStudentById = async (studentId, signup_id) => {
  const query = `
    SELECT s.*
    FROM students s
    JOIN signup u ON s.signup_id = u.id
    WHERE s.id = $1
      AND u.school_id = (
        SELECT school_id FROM signup WHERE id = $2
      )
    LIMIT 1
  `;

  try {
    const result = await pool.query(query, [studentId, signup_id]);
    return result.rows[0];
  } catch (err) {
    console.error('Error fetching student by id:', err);
    throw err;
  }
};


export const updateStudent = async (studentId, studentData) => {
  const {
    student_name,
    date_of_birth,
    gender,
    address,
    father_name,
    mother_name,
    assigned_class,
    assigned_section,
    birth_certificate,
    student_photo
  } = studentData;

  const updateQuery = `
    UPDATE students SET
      student_name = $1, 
      date_of_birth = $2, 
      gender = $3, 
      address = $4, 
      father_name = $5, 
      mother_name = $6,
      assigned_class = $7, 
      assigned_section = $8, 
      birth_certificate = $9, 
      student_photo = $10
    WHERE id = $11
    RETURNING *
  `;

  try {
    const result = await pool.query(updateQuery, [
      student_name,
      date_of_birth,
      gender,
      address,
      father_name,
      mother_name,
      assigned_class,
      assigned_section,
      birth_certificate,
      student_photo,
      studentId
    ]);
    return result;
  } catch (err) {
    console.error('Error updating student:', err);
    throw err;
  }
};
export const deleteStudent = async (studentId) => {
  const deleteStudentQuery = 'DELETE FROM students WHERE id = $1 RETURNING *';
  try {
    const result = await pool.query(deleteStudentQuery, [studentId]);
    if (result.rowCount === 0) {
      // Koi student nahi mila delete karne ke liye
      return null;
    }
    return result.rows[0];  // deleted student ka record return karo
  } catch (err) {
    console.error('Error deleting student:', err);
    throw err;
  }
};


// export const getStudentsByClass = async (className, signup_id) => {
//   const query = 'SELECT * FROM students WHERE assigned_class = $1 AND signup_id = $2';
//   try {
//     const result = await pool.query(query, [className, signup_id]);
//     return result.rows;
//   } catch (err) {
//     console.error('Error fetching students by class:', err);
//     throw err;
//   }
// };

// export const getStudentsByClassInSchool = async (className, signup_id) => {
//   const query = `
//     SELECT s.*
//     FROM students s
//     JOIN signup u ON s.signup_id = u.id
//     WHERE u.school_id = (
//       SELECT school_id FROM signup WHERE id = $1
//     )
//     AND LOWER(TRIM(s.assigned_class)) = LOWER(TRIM($2))
//   `;
//   try {
//     const result = await pool.query(query, [signup_id, className]);
//     return result.rows;
//   } catch (err) {
//     console.error('Error fetching students by class in school:', err);
//     throw err;
//   }
// };


export const getStudentsByClassInSchool = async (className, signup_id) => {
  const query = `
    SELECT 
      s.*,
      u.email AS username,
      u.password
    FROM students s
    JOIN signup u ON s.signup_id = u.id
    WHERE u.school_id = (
      SELECT school_id FROM signup WHERE id = $1
    )
    AND LOWER(TRIM(s.assigned_class)) = LOWER(TRIM($2))
  `;
  try {
    const result = await pool.query(query, [signup_id, className]);
    return result.rows;
  } catch (err) {
    console.error('Error fetching students by class in school:', err);
    throw err;
  }
};


export const getStudentsBySchoolId = async (signup_id) => {
  const query = `
    SELECT s.*
    FROM students s
    JOIN signup u ON s.signup_id = u.id
    WHERE u.school_id = (
      SELECT school_id FROM signup WHERE id = $1
    )
  `;
  try {
    const result = await pool.query(query, [signup_id]);
    return result.rows;
  } catch (err) {
    console.error('Error fetching students by school_id:', err);
    throw err;
  }
};


export const getStudentCount = async (signup_id) => {
  // const query = 'SELECT COUNT(*) AS totalstudents FROM students WHERE signup_id = $1';
  const query = `
  SELECT COUNT(*) AS totalstudents 
  FROM students 
  WHERE signup_id IN (
    SELECT id FROM signup WHERE school_id = (
      SELECT school_id FROM signup WHERE id = $1
    )
  )
`;

  const values = [signup_id];

  try {
    const result = await pool.query(query, values);
    return result.rows; // returns array like: [{ totalstudents: '3' }]
  } catch (err) {
    console.error('Database error in getStudentCount:', err);
    throw err;
  }
};
export const getStudentCountByClass = async (signup_id) => {
  const query = `
    SELECT
      LOWER(TRIM(assigned_class)) AS class_name,
      LOWER(TRIM(assigned_section)) AS section,
      COUNT(*) AS student_count
    FROM
      students
    WHERE signup_id IN (
  SELECT id FROM signup WHERE school_id = (
    SELECT school_id FROM signup WHERE id = $1
  )
)

      AND assigned_class IS NOT NULL
      AND assigned_class != ''
      AND assigned_section IS NOT NULL
      AND assigned_section != ''
    GROUP BY
      LOWER(TRIM(assigned_class)),
      LOWER(TRIM(assigned_section))
    ORDER BY
      LOWER(TRIM(assigned_class)),
      LOWER(TRIM(assigned_section))
  `;
  
  try {
    const result = await pool.query(query, [signup_id]);
    return result.rows;
  } catch (err) {
    console.error('Error getting student count by class:', err);
    throw err;
  }
};

export const getLastRegistrationNumber = async (signup_id) => {
  const query = `
    SELECT MAX(registration_number) AS lastregnumber
    FROM students
    WHERE signup_id IN (
  SELECT id FROM signup WHERE school_id = (
    SELECT school_id FROM signup WHERE id = $1
  )
)
  `;

  try {
    const result = await pool.query(query, [signup_id]);
    return result.rows[0]?.lastregnumber;
  } catch (err) {
    console.error('Error getting last registration number:', err);
    throw err;
  }
};

File: mobile/sms/lib/models/student_model.dart
// // class Student {
// //   final String id;
// //   final String name;
// //   final String registrationNumber;
// //   final String className;
// //   final String assignedSection;
// //   final String studentPhoto;
// //   final DateTime admissionDate;
// //   final String username;
// //   final String password;

// //   const Student({
// //     required this.id,
// //     required this.name,
// //     required this.registrationNumber,
// //     required this.className,
// //     required this.assignedSection,
// //     required this.studentPhoto,
// // required this.admissionDate,
// // required this.username,
// // required this.password,
// //   });
// // }

// // student_model.dart
// class Student {
//   String id;
//   String name;
//   String registrationNumber;
//   String dateOfBirth;
//   String gender;
//   String address;
//   String fatherName;
//   String motherName;
//   String email;
//   String phone;
//   String assignedClass;
//   String assignedSection;
//   String birthCertificate;
//   String studentPhoto;
//   DateTime admissionDate;
//   String username;
//   String password;
//   String createdAt;

//   Student({
//     required this.id,
//     required this.name,
//     required this.registrationNumber,
//     required this.dateOfBirth,
//     required this.gender,
//     required this.address,
//     required this.fatherName,
//     required this.motherName,
//     required this.email,
//     required this.phone,
//     required this.assignedClass,
//     required this.assignedSection,
//     required this.birthCertificate,
//     required this.studentPhoto,
//     required this.admissionDate,
//     required this.username,
//     required this.password,
//     required this.createdAt,
//   });

//   factory Student.fromJson(Map<String, dynamic> json) {
//     final photoData = json['student_photo'];
//     final birthCertData = json['birth_certificate'];

//     String processPath(dynamic data) {
//       if (data == null) return '';
//       if (data is Map && data['type'] == 'Buffer' && data['data'] is List) {
//         return String.fromCharCodes(List<int>.from(data['data']));
//       }
//       String path = data.toString();
//       if (path.contains('uploads\\')) {
//         return path.split('uploads\\').last;
//       }
//       return path;
//     }

//     return Student(
//       id: json['id']?.toString() ?? '',
//       name: json['student_name']?.toString() ?? 'Unknown',
//       registrationNumber: json['registration_number']?.toString() ?? '',
//       dateOfBirth: json['date_of_birth']?.toString() ?? '',
//       gender: json['gender']?.toString() ?? '',
//       address: json['address']?.toString() ?? '',
//       fatherName: json['father_name']?.toString() ?? '',
//       motherName: json['mother_name']?.toString() ?? '',
//       email: json['email']?.toString() ?? '',
//       phone: json['phone']?.toString() ?? '',
//       assignedClass: json['assigned_class']?.toString() ?? '',
//       assignedSection: json['assigned_section']?.toString() ?? '',
//       studentPhoto: processPath(photoData),
//       admissionDate: json['created_at'] != null
//           ? DateTime.parse(json['created_at'].toString())
//           : DateTime.now(),
//       username: json['username']?.toString() ?? 'N/A',
//       password: json['password']?.toString() ?? 'N/A',
//       birthCertificate: processPath(birthCertData),
//       createdAt: json['created_at']?.toString() ?? '',
//     );
//   }
// }

class Student {
  String id;
  String name;
  String registrationNumber;
  String dateOfBirth;
  String gender;
  String address;
  String fatherName;
  String motherName;
  String email;
  String phone;
  String assignedClass;
  String assignedSection;
  String birthCertificate;
  String studentPhoto;
  DateTime admissionDate;
  String username;
  String password;
  String createdAt;
  bool isPresent;

  Student({
    required this.id,
    required this.name,
    required this.registrationNumber,
    required this.dateOfBirth,
    required this.gender,
    required this.address,
    required this.fatherName,
    required this.motherName,
    required this.email,
    required this.phone,
    required this.assignedClass,
    required this.assignedSection,
    required this.birthCertificate,
    required this.studentPhoto,
    required this.admissionDate,
    required this.username,
    required this.password,
    required this.createdAt,
    this.isPresent = false,
  });

  factory Student.fromJson(Map<String, dynamic> json) {
    final photoData = json['student_photo'];
    final birthCertData = json['birth_certificate'];

    String processPath(dynamic data) {
      if (data == null) return '';
      if (data is Map && data['type'] == 'Buffer' && data['data'] is List) {
        return String.fromCharCodes(List<int>.from(data['data']));
      }
      String path = data.toString();
      if (path.contains('uploads\\')) {
        return path.split('uploads\\').last;
      }
      return path;
    }

    return Student(
      id: json['id']?.toString() ?? json['_id']?.toString() ?? '',
      name: json['student_name']?.toString() ?? 'Unknown',
      registrationNumber: json['registration_number']?.toString() ?? '',
      dateOfBirth: json['date_of_birth']?.toString() ?? '',
      gender: json['gender']?.toString() ?? '',
      address: json['address']?.toString() ?? '',
      fatherName: json['father_name']?.toString() ?? '',
      motherName: json['mother_name']?.toString() ?? '',
      email: json['email']?.toString() ?? '',
      phone: json['phone']?.toString() ?? '',
      assignedClass: json['assigned_class']?.toString() ??
          json['class_name']?.toString() ??
          '',
      assignedSection: json['assigned_section']?.toString() ??
          json['section']?.toString() ??
          '',
      studentPhoto: processPath(photoData),
      admissionDate: json['created_at'] != null
          ? DateTime.parse(json['created_at'].toString())
          : DateTime.now(),
      username: json['username']?.toString() ?? 'N/A',
      password: json['password']?.toString() ?? 'N/A',
      birthCertificate: processPath(birthCertData),
      createdAt: json['created_at']?.toString() ?? '',
    );
  }
}
